### PR TITLE
 #19398 fix stream leak in ProtocolSwitchStage

### PR DESCRIPTION
**Pull request by rkuhn**
also fix potential NPE in TCP streams when failed or canceled early for
an outgoing connection

### CODE REVIEW

**Comment by rkuhn** [comment]
Wow, the compilation error on Jenkins was not reproducible locally—how can scalac be different between the two? Anyway, erroneous import has been removed now.

**Comment by rkuhn** [comment]
fixes #19398 

/cc @jrudolph

File: akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala [hunk 1]
```diff
@@ -564,5 +566,8 @@ private[stream] class ActorGraphInterpreter(_initial: GraphInterpreterShell) ext
       }
   }
 
-  override def postStop(): Unit = activeInterpreters.foreach(_.tryAbort(AbruptTerminationException(self)))
+  override def postStop(): Unit = {
+    activeInterpreters.foreach(_.tryAbort(AbruptTerminationException(self)))
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala]: cache allocation?

File: akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala [hunk 2]
```diff
@@ -102,6 +102,89 @@ object GraphStages {
   private val _detacher = new Detacher[Any]
   def detacher[T]: GraphStage[FlowShape[T, T]] = _detacher.asInstanceOf[GraphStage[FlowShape[T, T]]]
 
+  class Breaker(callback: Breaker.Operation ⇒ Unit) {
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala]: final?

File: akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala (hunk 1 above)
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala]: I have no clue why this commit is in here, this should not be in this PR—will fix.

File: akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala (hunk 2 above)
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala]: yes

File: akka-http-core/src/test/scala/akka/http/impl/engine/ws/WebsocketIntegrationSpec.scala [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+  implicit val materializer = ActorMaterializer()
+
+  "A Websocket server" must {
+
+    "echo 100 elements and then shut down without error" in Utils.assertAllStagesStopped {
+      val bindingFuture = Http().bindAndHandleSync({
+        case HttpRequest(_, _, headers, _, _) ⇒
+          val upgrade = headers.collectFirst { case u: UpgradeToWebsocket ⇒ u }.get
+          upgrade.handleMessages(Flow.apply, None)
+      }, interface = "localhost", port = 8080)
```
Comment (ktoso) [inline akka-http-core/src/test/scala/akka/http/impl/engine/ws/WebsocketIntegrationSpec.scala]: `akka.testkit.SocketUtil.temporaryServerAddress.getPort` would be better

File: akka-http-core/src/test/scala/akka/http/impl/engine/ws/WebsocketIntegrationSpec.scala [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        Flow.fromSinkAndSourceMat(
+          Sink.fold(0)((n, _: Message) ⇒ n + 1),
+          Source.repeat(TextMessage("hello")).take(N))(Keep.left))
+
+      count.futureValue should ===(N)
+      binding.unbind()
+    }
+
+    "send back 100 elements and then terminate without error even when not ordinarily closed" in Utils.assertAllStagesStopped {
+      val N = 100
```
Comment (ktoso) [inline akka-http-core/src/test/scala/akka/http/impl/engine/ws/WebsocketIntegrationSpec.scala]: detail: could move N out of test and use in test message `s"send back #N elements..."`

**Comment by ktoso** [comment]
LGTM, thanks @rkuhn for the fix and @jrudolph for the diagnosis!

**Pull request merged**